src/webserver/DOSGuard.h

```diff
@@ -40,12 +50,110 @@ class BaseDOSGuard
 };
 
+class Whitelist
+{
+    std::vector<ip::network_v4> subnetsV4_;
+    std::vector<ip::network_v6> subnetsV6_;
+    std::vector<ip::address> ips_;
+
+    static bool
+    isInV4Subnet(ip::address const& addr, ip::network_v4 const& subnet)
+    {
+        auto const range = subnet.hosts();
+        return range.find(addr.to_v4()) != range.end();
+    }
+
+    static bool
+    isInV6Subnet(ip::address const& addr, ip::network_v6 const& subnet)
+    {
+        auto const range = subnet.hosts();
+        return range.find(addr.to_v6()) != range.end();
+    }
+
+public:
+    // could throw, handle with care
+    void
+    add(std::string_view net)
+    {
+        if (not isMask(net))
+        {
+            ips_.push_back(ip::make_address(net));
+            return;
+        }
+
+        if (isV4(net))
+            subnetsV4_.push_back(ip::make_network_v4(net));
+        else if (isV6(net))
+            subnetsV6_.push_back(ip::make_network_v6(net));
+        else
+            throw std::runtime_error(std::string{"malformed network: "} + net.data());
+    }
+
+    // could throw, handle with care
+    bool
+    isWhiteListed(std::string_view ip) const
+    {
+        auto const addr = ip::make_address(ip);
+        if (std::find(std::begin(ips_), std::end(ips_), addr) != std::end(ips_))
+            return true;
+
+        if (addr.is_v4())
+            return std::find_if(
+                       std::begin(subnetsV4_), std::end(subnetsV4_), std::bind_front(&isInV4Subnet, std::cref(addr))) !=
+                std::end(subnetsV4_);
+
+        if (addr.is_v6())
+            return std::find_if(
+                       std::begin(subnetsV6_), std::end(subnetsV6_), std::bind_front(&isInV6Subnet, std::cref(addr))) !=
+                std::end(subnetsV6_);
+
+        return false;
+    }
+
+private:
+    bool
+    isV4(std::string_view net) const
+    {
+        static const std::regex ipv4CidrRegex(R"(^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/\d{1,2}$)");
+        return std::regex_match(std::string(net), ipv4CidrRegex);
+    }
+    bool
+    isV6(std::string_view net) const
+    {
+        static const std::regex ipv6CidrRegex(R"(^([0-9A-Fa-f]{1,4}:){7}[0-9A-Fa-f]{1,4}/\d{1,3}$)");
+        return std::regex_match(std::string(net), ipv6CidrRegex);
+    }
+    bool
+    isMask(std::string_view net) const
+    {
+        return net.find('/') != std::string_view::npos;
+    }
+};
+
+class WhitelistHandler
+{
+    Whitelist whitelist_;
+
+public:
+    WhitelistHandler(std::unordered_set<std::string> const& arr)
+    {
+        for (auto const& net : arr)
+            whitelist_.add(net);
+    }
+
+    bool
+    isWhiteListed(std::string_view ip) const
+    {
+        return whitelist_.isWhiteListed(ip);
+    }
+};
+
 /**
  * @brief A simple denial of service guard used for rate limiting.
  *
  * @tparam SweepHandler Type of the sweep handler
  */
-template <typename SweepHandler>
+template <typename WhitelistHandlerType, typename SweepHandler>
```

Comment:
Feel free to also change `SweepHandler` to `SweepHandlerType` to match the other one.